### DISCUSSION TITLE
OpenAPI Client Generators with support 3.1.0 specs

### DISCUSSION THREAD

**Discussion by quizmoon** [fastapi/fastapi#9810] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
Openapi	"3.1.0"
```

### Description

FastAPI 0.99 came upgrade of OpenAPI to 3.1.0 

So far I could not find any OpenAPI Client Generators which have a announced support of 3.1.0 specs. 

If somebody knows one would be great to know also. 

Thank you!

### Operating System

Linux

### FastAPI Version

0.99.1

### Python Version

3.11

**Comment by gnir-work** [comment]
Hey,
This issue is also affecting us, as we heavily rely on [openapi-generator](https://openapi-generator.tech/).
I tried referencing the documentation of `fastapi` in hope to find a configuration related to the version of the `open api` spec used. 
I found the `openapi_version` parameter however it only changed the version in the generated spec but still used the new features only supported in `3.1.0`.

In addition to an alternative tool a way to tell `fastapi` to generate an older version spec will also help us a lot!
Thanks in advance to any answers and guidance.

**Comment by sklnd** [comment]
I had this same problem after upgrading FastAPI yesterday. 

[kubb](https://www.kubb.dev/) is listed on the [OpenAPI Code Generators](https://tools.openapis.org/categories/code-generators.html) list as supporting OpenAPI 3.1.

 It's a bit of a rewrite to rip out OpenAPI Generator, but so far I am finding it serviceable after turning on a few plugins.

**Comment by commonism** [comment]
Not a generator but a dynamic client - [aiopenapi3](https://aiopenapi3.readthedocs.io/en/latest/) does OpenAPI3 v3.1.
You may want to use the [to-be-merged](https://github.com/commonism/aiopenapi3/pull/95) pydanticv2 branch when giving it a shot.

**Reply by abrahan-munro** [reply]
That looks like a nice approach! I'd like to try it when I have time.

I guess type hints would be missing. Are you planning on generating stubs?

Also, I'm using FastAPI rather than writing the document by hand, so I'm not sure if any of the listed limitations would be an issue.

**Comment by abrahan-munro** [comment]
A warning should be added to https://fastapi.tiangolo.com/advanced/generate-clients/, something like:

> **Warning**
> FastAPI supports only OpenAPI 3.1 but almost all code generators, including those mentioned on this page, support only OpenAPI 3.0.x, and so are **not compatible**. For code generators supporting OpenAPI 3.1 and hence FastAPI, see https://tools.openapis.org/categories/code-generators.html and note the *3.1?* column. At the time of writing, the only compatible generator is https://www.kubb.dev/, and that supports only TypeScript.

**Comment by abrahan-munro** [comment]
What are the options for code generation with FastAPI?
- Is there any way to have FastAPI generate 3.0.x?
- Is there any way to externally convert from 3.1 to 3.0.x?

**Reply by quizmoon** [reply]
There is a trick, please read more here: https://github.com/tiangolo/fastapi/discussions/9789#discussioncomment-6355603

**Reply by abrahan-munro** [reply]
Thanks! That turned out to be the solution for me, with OpenAPITools/openapi-generator generating a python client.

**Comment by abrahan-munro** [comment]
I'm writing a python client for use with FastAPI. Of the available python client generators, https://github.com/openapi-json-schema-tools/openapi-json-schema-generator by @spacether  seems to be the only one that might be actively working on supporting OpenAPI 3.1.

What are the alternatives to code generation for python? Just do it manually with `requests` and `pydantic` and copy the schema files from the server source?

**Reply by quizmoon** [reply]
@spacether created that project https://github.com/orgs/OpenAPITools/projects/4 to add support of 3.1, is not so close yet to the end, unfortunately.

**Reply by commonism** [reply]
IMHO dynamic clients suit the fluid nature of the description document due to the code-first approach of FastAPI even better than generating a static client, but I may be biased.
Current FastAPI uses pydantic v2, pydantic v2 generated Json Schemas are not compatible with OpenAPI v3.0.
The lack of support for OpenAPI v3.1 is a problem with the current eco-system, redoc/swagger-ui will haunt you as well.
I'm glad there is pioneers taking inconvenient but necessary steps. OpenAPI v3.1 is about [2 years old](https://github.com/OAI/OpenAPI-Specification/commits/main/versions/3.1.0.md) already and having FastAPI use it already has an positive impact on the eco-system.
pydantic/FastAPI even only use a very small subset of the Json Schema specification which is the foundation of OpenAPI v3.1.

**Reply by quizmoon** [reply]
On dynamic clients you should always have schema available, local or remotely, it is also raise question of performance, as every time you want to use it - client should parse schema again and again. 

So far, there are no options, except using trick with  override of openapi_version and using pydantic v1(as v2 schema is not compatible with OpenAPI v3.0)

**Reply by abrahan-munro** [reply]
https://github.com/mom1/api-client-pydantic by 
@mom1 also looks promising, although no pydantic2 support yet.

**Reply by spacether** [reply]
https://github.com/openapi-json-schema-tools/openapi-json-schema-generator now has > 50% of the new 3.1.0 json schema keyword features added in the latest commit. The project tracking the features is [here](https://github.com/orgs/openapi-json-schema-tools/projects/4)

**Comment by commonism** [comment]
https://github.com/mom1/api-client-pydantic is not a client in the sense you feed it the description document but a client in the sense you get a thin layer of assistance in implementing the protocol.

**Reply by abrahan-munro** [reply]
Yes, https://github.com/tiangolo/fastapi/discussions/9810#discussioncomment-6621894 is discussing alternatives to code generation.

**Reply by commonism** [reply]
Should have been a reponse to https://github.com/tiangolo/fastapi/discussions/9810#discussioncomment-6632592 - I'm not that comfortable with the  discussion UI yet and fail to write to the proper textfield.

**Reply by abrahan-munro** [reply]
That's no problem. I just meant yes api-client-pydantic isn't a client generator, but can help in writing clients in the absence of a working generator, so might be a viable alternative. For python clients, it turns out that openapi-generator can be used with current FastAPI after all: https://github.com/tiangolo/fastapi/discussions/9810#discussioncomment-6634308.

**Reply by commonism** [reply]
Generating a client by force feeding a v3.1 description document in a generator by masking it as v3.0 does not sound reasonable to me.
You can't even have "null" properties - v3.1 that is `type: […, null]` - in v3.0 `nullable: True`.
It's even bad for the generator to even accept v3.1 language in a description document which claims to be v3.0.

Therefore I'd expect `mixed results` and would not recommend.

**Reply by abrahan-munro** [reply]
It's certainly not a general solution, and I'd much rather use a generator with 3.1.0 support, but in the case of openapi-generator with FastAPI, it seems to work ok (and type: null seems to be handled): https://github.com/tiangolo/fastapi/discussions/9789#discussioncomment-6361207.

**Comment by feliche93** [comment]
Anyone looking for a typescript fetch client generator, this worked for me well so far: https://openapi-ts.pages.dev/

**Comment by YuriiMotov** [comment]
https://fastapi.tiangolo.com/advanced/generate-clients/
